core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -149,12 +153,23 @@ class KafkaApis(val requestChannel: RequestChannel,
       }
     }
 
-    forwardingManager match {
-      case Some(mgr) if !request.isForwarded && !controller.isActive =>
-        mgr.forwardRequest(request, responseCallback)
+    metadataSupport match {
+      // ZooKeeper
+      case zkSupport@ZkSupport(_, _, _, forwardingManager) =>
+        forwardingManager match {
+          case Some(mgr) if !request.isForwarded && !zkSupport.controller.isActive =>
+            mgr.forwardRequest(request, responseCallback)
 
-      case _ =>
-        handler(request)
+          case _ =>
+            handler(request)
+        }
+      // Raft
+      case RaftSupport(fwdMgr) =>
+        if (!request.isForwarded) {
+          fwdMgr.forwardRequest(request, responseCallback)
+        } else {
+          handler(request) // will reject
+        }
```

Comment:
One way to move this complexity to the implementations of `MetadataSupport` is by adding the following method:

```scala
def forwardRequest(request: RequestChannel.Request, callback: Option[AbstractResponse] => Unit): Boolean
```

With this function, here you can:

```scala
if (!metadataSupport.forwardRequest(request, responseCallback)) {
  handler(request);
}
```

Reply:
I added a maybeForward() method to the trait